src/h3/stream.rs

```diff
@@ -370,51 +332,130 @@ impl Stream {
+        );
+
+        self.state_off += read;
+
+        if !self.state_buffer_complete() {
+            return Err(Error::Done);
+        }
+
+        Ok(())
+    }
+
+    /// Tries to fill the state buffer by reading data from the given cursor.
+    ///
+    /// This is intended to replace `try_fill_buffer()` in tests, in order to
+    /// avoid having to setup a transport connection.
+    #[cfg(test)]
+    fn try_fill_buffer_for_tests(
+        &mut self, stream: &mut std::io::Cursor<Vec<u8>>,
+    ) -> Result<()> {
+        let buf = &mut self.state_buf[self.state_off..self.state_len];
+
+        let read = std::io::Read::read(stream, buf).unwrap();
+
+        self.state_off += read;
+
+        if !self.state_buffer_complete() {
+            return Err(Error::Done);
+        }
+
+        Ok(())
+    }
+
+    /// Tries to parse a varint (including length) from the state buffer.
+    pub fn try_consume_varint(&mut self) -> Result<u64> {
+        if self.state_off == 1 {
+            self.state_len = octets::varint_parse_len(self.state_buf[0]);
+            self.state_buf.resize(self.state_len, 0);
+        }
+
+        // Return early if we don't have enough data in the state buffer to
+        // parse the whole varint.
+        if !self.state_buffer_complete() {
+            return Err(Error::Done);
+        }
+
+        let varint =
+            octets::Octets::with_slice(&mut self.state_buf).get_varint()?;
+
+        Ok(varint)
+    }
+
+    /// Tries to parse a frame from the state buffer.
+    pub fn try_consume_frame(&mut self) -> Result<()> {
+        // TODO: properly propagate frame parsing errors.
         if let Ok(frame) = frame::Frame::from_bytes(
             self.frame_type.unwrap(),
-            self.frame_payload_len,
-            self.buf_bytes(self.frame_payload_len as usize)?,
+            self.state_len as u64,
+            &mut self.state_buf,
         ) {
             self.frames.push_back(frame);
         }
 
-        // TODO: bytes in the buffer are no longer needed, so we can remove
-        // them and set the offset back to 0?
-        self.buf_read_off += self.frame_payload_len;
-
-        // Stream offset always increases, so we can track how many total
-        // bytes were seen by the application layer.
-        self.stream_offset += self.frame_payload_len;
-
-        // Set state to parse next frame
-        self.state = State::FrameTypeLen;
+        self.state_transition(State::FrameType, 1);
 
         Ok(())
     }
 
-    pub fn more(&self) -> bool {
-        let rem_bytes = self.buf_end_pos - self.buf_read_off;
-        rem_bytes > 0
+    /// Returns true if the state buffer has enough data to complete the state.
+    fn state_buffer_complete(&self) -> bool {
+        self.state_off == self.state_len
+    }
+
+    /// Transitions the stream to a new state, and resets the state buffer.
+    fn state_transition(&mut self, new_state: State, expected_len: usize) {
+        self.state = new_state;
```

Comment:
At some point in the distant past we had a function like. It was a convenient place to put transition sanity checking. We removed it though for a range of reasons.
We could add that back, however, given that all state transitions are handled in internally to stream.rs, the overhead of checks might be unnecessary.

Reply:
Well, this function is really just to avoid having to copy the same 3 lines all over the place. We already have asserts (which are already of dubious utility) to validate state transitions, so more runtime checks are unnecessary IMO.

What we might want to do is some form of static checking like the one proposed in https://hoverbear.org/2016/10/12/rust-state-machine-pattern/ but I'm not convinced it would improve the situation much TBH.

Reply:
The changes that would be needed for static checking seem quite onerous.

We have a bunch of tests that validate transitions. So on reflection I think what we have is ok.